### ISSUE TITLE
Memory allocation with column-length-limit

### ISSUE THREAD

**Issue by ErikJansenIRefact** [pacman82/odbc2parquet#456]
Hi Markus,

So far the odbc2parquet works fast and stable (we are very pleased with the utility).
But we now need to export a table in SQLServer with varchar(max) and nvarchar(max) and varbinary(max) columns.
If we do not specify the column-length-limit switch these columns will not be exported. That's not what our user expects.

So we set the column-length-limit to 32.000 (max this amount of bytes will be exported). During the export we do get some odbc truncation errors which we will simply ignore.

But we now also see huge memory allocations. If we set batch-size-row to 1000000 in combination with column-length-limit to for example 8000 more than 15Gbyte of memory is allocated to the odbc2parquet. Even if the table itself only has 100 rows.

Can you explain how the memory allocation works so we can use the settings wisely?

Upfront we are not aware of how many rows a table contains (might be only a few rows are a huge amount of rows).

**Comment by pacman82** [comment]
`odbc2parquet` is fast because it uses preallocated memory and bulk inserts. The actual amount of rows in the table does matter in the least for memory consumption only the schema of the table does. If you set `--batch-size-row` to 1.000.000 it will allocate enough memory for a million rows of that schema. If you enable verbose logging you should be able to see how much each column requires per row.

Here is the `--help` text of what `--batch-size-row` does:

`--batch-size-row <BATCH_SIZE_ROW>`

> Size of a single batch in rows. The content of the data source is written into the output parquet files in batches. This way the content does never need to be materialized completely in memory at once. If `--batch-size-memory` is not specified this value defaults to 65535. This avoids issues with some ODBC drivers using 16Bit integers to represent batch sizes. If `--batch-size-memory` is specified no other limit is applied by default. If both option are specified the batch size is the largest possible which satisfies both constraints

If you do not set anything at all. The `odbc2parquet` will choose the batch size itself, so it can fit neatly within 2 GiB. If you aim for a different utilization of memory you can set `batch-size-memory`.

`--batch-size-memory`

> Limits the size of a single batch. It does so by calculating the amount of memory each row requires in the allocated buffers and then limits the maximum number of rows so that the maximum buffer size comes as close as possible, but does not exceed the specified amount. Default is 2GiB on 64 Bit platforms and 1GiB on 32 Bit Platforms if `--batch-size-row` is not specified. If `--batch-size-row` is not specified no memory limit is applied by default. If both option are specified the batch size is the largest possible which satisfies both constraints. This option controls the size of the buffers of data in transit, and therefore the memory usage of this tool. It indirectly controls the size of the row groups written to parquet (since each batch is written as one row group). It is hard to make a generic statement about how much smaller the average row group will be. This options allows you to specify the memory usage using SI units. So you can pass `2Gib`, `600Mb` and so on

Best, Markus

**Comment by pacman82** [comment]
tl;dr I would recommend just to remove the `--batch-size-row` flag and its argument.

**Comment by pacman82** [comment]
```rust
impl BatchSizeLimit {
    pub fn new(num_rows_limit: Option<usize>, memory_limit: Option<ByteSize>) -> Self {
        match (num_rows_limit, memory_limit) {
            (Some(rows), None) => BatchSizeLimit::Rows(rows),
            (None, Some(memory)) => BatchSizeLimit::Bytes(memory),
            // User specified nothing => Use default
            (None, None) => BatchSizeLimit::Both {
                rows: DEFAULT_BATCH_SIZE_ROWS,
                memory: DEFAULT_BATCH_SIZE_BYTES,
            },
            (Some(rows), Some(memory)) => BatchSizeLimit::Both { rows, memory },
        }
    }
  ```
  
  There are two limits rows and memory. If you set one explicitly the other is deactivated unless you set both explicitly. Defaults for row limit is `65535` default for memory limit is `2 GiB`.

**Comment by ErikJansenIRefact** [comment]
Hi Markus,

We've played around a little bit with the different settings and we now see how it works. Thanks for your quick response.

Note: there is one potential improvement in the code. If you set column-length-limit this limit seems to be applied on all text like columns. But for many columns the size limit is smaller than what is probably set in column-length-limit leading to a waste of allocated memory.

For example: assume a table A(col1 varchar(10), col2 varchar(10), col3 varchar(max)). And you set column-length-limit to 100000. Then for col1 and col2 the mem allocation is 100000 instead of 10. In this case the odbc driver reports a max size of 10 so no need to use column-length-limit in this case.

Do you think an issue should be created for this?

**Comment by pacman82** [comment]
Hello @ErikJansenIRefact ,

oh dear. 😱 You catched an honest to goodness Bug. I would be all for an issue, but this one is fast to fix so `odbc2parquet 2.0.4` is already released which only applies the limit to large columns.

**Comment by pacman82** [comment]
Having a user of the tool at hand. Would you prefer the tool to error out if the driver reports a column with size `0`?

**Comment by ErikJansenIRefact** [comment]
We often have varchar(max), nvarchar(max) columns. So with the column-length-limit option we now do have the option to export the data. We are also able to capture the warning messages. So no real need to error out.

Perhaps as a flag?

**Comment by pacman82** [comment]
>  So no real need to error out

This is precisely the question. Would error out be the better default?

**Comment by ErikJansenIRefact** [comment]
Yes it is. Because a warning might not be seen and your parquet file misses columns.

**Comment by pacman82** [comment]
Thank you. Likely I'll do a breaking change soon and release a version 3.0 in the future, which breaks by default. Would a migration to the new default behavior be a lot of effort on your side?

**Comment by ErikJansenIRefact** [comment]
No, as long as when specifying column-length-limit keeps working.
We check the table on max columns and apply a column-length-limit dynamically for these tables.
